### PR TITLE
Share string interner between scrape cache and remote_write

### PR DESCRIPTION
**Pull request by tpaschalis**
This PR is a follow-up on #7038.

It is an attempt to publicly expose the string interner that exists for remote_write, so its shared pool can also be used by the scrape cache, decreasing memory usage for labels appearing in both places. 

This is hidden behind a feature flag, like other experimental features. If you agree on this approach, I'll also update the documentation to mention the new flag.

As per the original PR, note that the string_interner_zero_reference_releases_total series now has a different namespace and subsystem from what it was initially set to in the RW code.

~Also note that this PR changes the seriesCur/SeriesPrev maps to use `*cacheEntry` as its value to reduce memory usage and allocations. In case you'd like to see this in a different PR to measure its effectiveness, I can revert this specific change and open it in a separate PR.~

cc @rfratto for more historical context.

### CODE REVIEW

**Comment by rfratto** [comment]
For some extra context, we've been running this code in Grafana Agent since the first release. IMO, now that Prometheus Agent has been released, there's renewed precedent to move the last bit of modified Prometheus code that we use upstream. 

The [prombench run][] from the previous PR didn't show any general improvement. The previous PR seems to be impacted by a bug that we've since fixed and added tests for (grafana#21). I don't expect that a new prombench run would be any different, with the exception of improved alloc/s given the bug we identified was over-interning. This PR would only demonstrate improved memory usage if remote_write is being used. 

I'd also like to point out that even with this being merged, Prometheus Agent still wouldn't be using it yet. Gut feeling is that Prometheus Agent can be updated to use the shared interner in a follow up PR. 

[prombench run]: http://prombench.prometheus.io/grafana/d/7gmLoNDmz/prombench?orgId=1&var-pr-number=7038&from=1599854857632&to=1600115345921

File: scrape/scrape.go [hunk 1]
```diff
@@ -882,14 +884,17 @@ type scrapeCache struct {
...
 
-	// seriesCur and seriesPrev store the labels of series that were seen
-	// in the current and previous scrape.
+	// seriesCur and seriesPrev store the ref of series that were seen
+	// in the current and previous scrape based on the hash.
 	// We hold two maps and swap them out to save allocations.
-	seriesCur  map[uint64]labels.Labels
-	seriesPrev map[uint64]labels.Labels
+	seriesCur  map[uint64]*cacheEntry
+	seriesPrev map[uint64]*cacheEntry
```
[the remark this answers is not in the document]
Comment (rfratto) [reply scrape/scrape.go:891]: It was a long time ago when I wrote this, so unfortunately I don't remember why I made this change. I _believe_ using a pointer helped me ensure that you wouldn't make extra copies of the labels when reading from the map, but I'm not completely confident.

File: model/intern/intern.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+// Shared interner
+var Global = New(prometheus.DefaultRegisterer)
```
Comment (cstyan) [inline model/intern/intern.go:31]: what reason does this need to be exported? I think we should either pass in the registerer used elsewhere in the code, or if one is not passed just default to use `prometheus.DefaultRegisterer`

**Review by cstyan — COMMENTED** [review]
@roidelapluie ping since this mostly impacts code you maintain

File: scrape/scrape.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpaschalis) [reply scrape/scrape.go:891]: Yeah, indeed it seems that using the pointer makes a difference in memory usage. I know that Prometheus as a whole is a much more complex system, but here are the differences when we run the included `BenchmarkStringInterner` where old is using `labels.Labels` as the map values, and new is using `*cacheEntry`.


For 1k metrics
```
name              old time/op    new time/op    delta
StringInterner-8    1.57ms ± 3%    1.11ms ± 0%  -29.19%  (p=0.000 n=10+9)

name              old alloc/op   new alloc/op   delta
StringInterner-8     734kB ± 0%     120kB ± 0%  -83.67%  (p=0.000 n=9+8)

name              old allocs/op  new allocs/op  delta
StringInterner-8     7.03k ± 0%     1.04k ± 0%  -85.17%  (p=0.000 n=9+8)
```

For 1M metrics
```
name              old time/op    new time/op    delta
StringInterner-8     1.90s ± 3%     2.17s ± 5%  +14.22%  (p=0.000 n=10+10)

name              old alloc/op   new alloc/op   delta
StringInterner-8     895MB ± 0%     469MB ± 0%  -47.61%  (p=0.000 n=10+10)

name              old allocs/op  new allocs/op  delta
StringInterner-8     7.21M ± 0%     3.05M ± 0%  -57.70%  (p=0.000 n=10+10)
```


Do you think that it's okay to keep this change, or would you prefer that we limit scope for this PR?

**Comment by tpaschalis** [comment]
@csmarchbanks I see, it makes sense to see how it ties together with the rest of Prometheus' moving pieces.

I've opened a draft PR on #10453 with the same code, but instances of the feature flag set to 'true'. Do you have permissions to kick off a prombench run? Do I have to do anything to enable it?

**Comment by roidelapluie** [comment]
We are looking at a 5% TSDB size increase with this change:

compaction


I can not really explain why.

**Comment by rfratto** [comment]
@roidelapluie is that TSDB disk size or head chunk memory size?

**Comment by rfratto** [comment]
@roidelapluie the benchmark results come from #10460, right? Does that run with the new feature flag enabled?

File: scrape/scrape.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpaschalis) [reply scrape/scrape.go:897]: I'll make an attempt to get the ball on this PR rolling once more.

I'm going to revert the use of `*cacheEntry` as the map key, to facilitate reviewing and have a smaller change radius. We can open and test the effect of this optimization in a separate PR.

**Comment by tpaschalis** [comment]
Hey @roidelapluie! I'd like to try and bring this back from the dead after some time.

Do you remember how you ended up with the 5% TSDB size increase estimate? The prombench data is available [here](http://prombench.prometheus.io/grafana/d/7gmLoNDmz/prombench?orgId=1&var-pr-number=10453&from=1647380506392&to=1647436920647); I'm asking so that I could replicate this effect and figure out how we could amend it.

**Comment by roidelapluie** [comment]
The data was comming from Prombench, First Compaction, Avg Bytes/Sample.

**Comment by bboreham** [comment]
Note this is will achieve nothing when Prometheus is built with  the “stringlabels” tag. If we can get #12304 to work that should be far better.

**Comment by tpaschalis** [comment]
Since we're no longer utilizing this in Grafana Agent to gauge its merit, and its usefulness has been superseded by other work, I'll close this PR.

Thanks!

**Pull request closed without merging**